### PR TITLE
OpenVX HIP GPU Backend - Fix for vxReplicateNode

### PR DESCRIPTION
**Pull request by hansely**
This vxReplicateNode fix has revealed hidden errors in the GPU flow that includes:
-vxWarpAffine
-vxCreateImageFromChannel
-vxLaplacianPyramid
-vxColorConvert

This PR includes the fixes for all the above errors (both OCL & HIP backend) except for vxLaplacianPyramid which is still in progress.

### CODE REVIEW

File: amd_openvx/openvx/api/vx_api.cpp [hunk 1]
```diff
@@ -944,13 +944,21 @@ VX_API_ENTRY vx_image VX_API_CALL vxCreateImageFromHandle(vx_context context, vx
                         data->children[i]->buffer = (vx_uint8 *)(ptrs ? ptrs[i] : nullptr);
                         data->children[i]->u.img.stride_in_bytes = addrs[i].stride_y;
                         data->children[i]->gpu_buffer_offset = 0;
+#if (ENABLE_OPENCL || ENABLE_HIP)
```
Comment (rrawther) [inline amd_openvx/openvx/api/vx_api.cpp:947]: Not sure why this is needed when creating the data

File: amd_openvx/openvx/api/vx_api.cpp [hunk 2]
```diff
@@ -944,13 +944,21 @@ VX_API_ENTRY vx_image VX_API_CALL vxCreateImageFromHandle(vx_context context, vx
...
                     }
                 }
                 else {
                     data->import_type = VX_MEMORY_TYPE_HOST;
                     data->buffer = (vx_uint8 *)(ptrs ? ptrs[0] : nullptr);
                     data->u.img.stride_in_bytes = addrs[0].stride_y;
                     data->gpu_buffer_offset = 0;
+#if (ENABLE_OPENCL || ENABLE_HIP)
+                    data->buffer_sync_flags &= ~AGO_BUFFER_SYNC_FLAG_DIRTY_MASK;
+                    data->buffer_sync_flags |= AGO_BUFFER_SYNC_FLAG_DIRTY_BY_COMMIT;
```
Comment (rrawther) [inline amd_openvx/openvx/api/vx_api.cpp:960]: Not sure why this is needed here

File: amd_openvx/openvx/api/vx_api.cpp (hunk 1 above)
Comment (LakshmiKumar23) [reply amd_openvx/openvx/api/vx_api.cpp:947]: @rrawther  for the createImageFromChannel test case I had to add this

File: amd_openvx/openvx/ago/ago_kernel_api.cpp [hunk 3]
```diff
@@ -2301,9 +2301,6 @@ int ovxKernel_LaplacianPyramid(AgoNode * node, AgoKernelCommand cmd)
     else if (cmd == ago_kernel_cmd_query_target_support) {
         node->target_support_flags = AGO_KERNEL_FLAG_SUBGRAPH
                     | AGO_KERNEL_FLAG_DEVICE_CPU
-#if ENABLE_OPENCL
```
Comment (rrawther) [inline amd_openvx/openvx/ago/ago_kernel_api.cpp:2304]: why is this removed? Are we supporting Laplacian Pyramid only on CPU?

Comment (LakshmiKumar23) [reply amd_openvx/openvx/ago/ago_kernel_api.cpp:2304]: @rrawther As of now yes, that's the whole reason for the test fail

**Review by rrawther — COMMENTED** [review]
Needs to review why we need to set the dirty flags at data creation time?

**Comment by hansely** [comment]
CI is passing now @asalmanp @kiritigowda

**Comment by kiritigowda** [comment]
PR verified - Final results of CTS below. Looks good to merge - @rrawther @asalmanp do you need any changes? Else, we can merge this.

* CTS - CPU Flow -- Pass

* CTS - GPU Flow
  + OpenCL - 
```
[ ======== ]
[ ALL DONE ] 5820 test(s) from 69 test case(s) ran
[ PASSED   ] 5811 test(s)
[ FAILED   ] 9 test(s), listed below:
[ FAILED   ] Graph.ImageContainmentRelationship
[ FAILED   ] LaplacianPyramid.GraphProcessing/0/randomInput/VX_BORDER_UNDEFINED/sz=128x128
[ FAILED   ] LaplacianPyramid.GraphProcessing/1/randomInput/VX_BORDER_UNDEFINED/sz=256x256
[ FAILED   ] LaplacianPyramid.GraphProcessing/2/randomInput/VX_BORDER_UNDEFINED/sz=640x480
[ FAILED   ] LaplacianPyramid.GraphProcessing/18/lena/VX_BORDER_UNDEFINED
[ FAILED   ] LaplacianPyramid.ImmediateProcessing/0/randomInput/VX_BORDER_UNDEFINED/sz=128x128
[ FAILED   ] LaplacianPyramid.ImmediateProcessing/1/randomInput/VX_BORDER_UNDEFINED/sz=256x256
[ FAILED   ] LaplacianPyramid.ImmediateProcessing/2/randomInput/VX_BORDER_UNDEFINED/sz=640x480
[ FAILED   ] LaplacianPyramid.ImmediateProcessing/18/lena/VX_BORDER_UNDEFINED
[ DISABLED ] 8190 test(s)
```
  + HIP - 
```
[ ======== ]
[ ALL DONE ] 5820 test(s) from 69 test case(s) ran
[ PASSED   ] 5796 test(s)
[ FAILED   ] 24 test(s), listed below:
[ FAILED   ] Graph.ImageContainmentRelationship
[ FAILED   ] vxMapRemapPatch.MapRandomRemap
[ FAILED   ] ColorConvert.OnRandomAndNatural/6/Immediate/RGB=>YUV4
[ FAILED   ] ColorConvert.OnRandomAndNatural/7/Graph/RGB=>YUV4
[ FAILED   ] ColorConvert.OnRandomAndNatural/14/Immediate/RGBX=>YUV4
[ FAILED   ] ColorConvert.OnRandomAndNatural/15/Graph/RGBX=>YUV4
[ FAILED   ] HarrisCorners.GraphProcessing/191/many_strong_corners/MIN_DISTANCE=30.0/SENSITIVITY=0.04/GRADIENT_SIZE=3/BLOCK_SIZE=7
[ FAILED   ] HarrisCorners.GraphProcessing/194/many_strong_corners/MIN_DISTANCE=30.0/SENSITIVITY=0.04/GRADIENT_SIZE=5/BLOCK_SIZE=7
[ FAILED   ] HarrisCorners.GraphProcessing/197/many_strong_corners/MIN_DISTANCE=30.0/SENSITIVITY=0.04/GRADIENT_SIZE=7/BLOCK_SIZE=7
[ FAILED   ] HarrisCorners.GraphProcessing/200/many_strong_corners/MIN_DISTANCE=30.0/SENSITIVITY=0.10/GRADIENT_SIZE=3/BLOCK_SIZE=7
[ FAILED   ] HarrisCorners.GraphProcessing/209/many_strong_corners/MIN_DISTANCE=30.0/SENSITIVITY=0.15/GRADIENT_SIZE=3/BLOCK_SIZE=7
[ FAILED   ] HarrisCorners.ImmediateProcessing/191/many_strong_corners/MIN_DISTANCE=30.0/SENSITIVITY=0.04/GRADIENT_SIZE=3/BLOCK_SIZE=7
[ FAILED   ] HarrisCorners.ImmediateProcessing/194/many_strong_corners/MIN_DISTANCE=30.0/SENSITIVITY=0.04/GRADIENT_SIZE=5/BLOCK_SIZE=7
[ FAILED   ] HarrisCorners.ImmediateProcessing/197/many_strong_corners/MIN_DISTANCE=30.0/SENSITIVITY=0.04/GRADIENT_SIZE=7/BLOCK_SIZE=7
[ FAILED   ] HarrisCorners.ImmediateProcessing/200/many_strong_corners/MIN_DISTANCE=30.0/SENSITIVITY=0.10/GRADIENT_SIZE=3/BLOCK_SIZE=7
[ FAILED   ] HarrisCorners.ImmediateProcessing/209/many_strong_corners/MIN_DISTANCE=30.0/SENSITIVITY=0.15/GRADIENT_SIZE=3/BLOCK_SIZE=7
[ FAILED   ] LaplacianPyramid.GraphProcessing/0/randomInput/VX_BORDER_UNDEFINED/sz=128x128
[ FAILED   ] LaplacianPyramid.GraphProcessing/1/randomInput/VX_BORDER_UNDEFINED/sz=256x256
[ FAILED   ] LaplacianPyramid.GraphProcessing/2/randomInput/VX_BORDER_UNDEFINED/sz=640x480
[ FAILED   ] LaplacianPyramid.GraphProcessing/18/lena/VX_BORDER_UNDEFINED
[ FAILED   ] LaplacianPyramid.ImmediateProcessing/0/randomInput/VX_BORDER_UNDEFINED/sz=128x128
[ FAILED   ] LaplacianPyramid.ImmediateProcessing/1/randomInput/VX_BORDER_UNDEFINED/sz=256x256
[ FAILED   ] LaplacianPyramid.ImmediateProcessing/2/randomInput/VX_BORDER_UNDEFINED/sz=640x480
[ FAILED   ] LaplacianPyramid.ImmediateProcessing/18/lena/VX_BORDER_UNDEFINED
[ DISABLED ] 8190 test(s)
```

**Comment by LakshmiKumar23** [comment]
> PR verified - Final results of CTS below. Looks good to merge - @rrawther @asalmanp do you need any changes? Else, we can merge this.
> 
> * CTS - CPU Flow -- Pass
> * CTS - GPU Flow
>   
>   * OpenCL -
> 
> ```
> [ ======== ]
> [ ALL DONE ] 5820 test(s) from 69 test case(s) ran
> [ PASSED   ] 5811 test(s)
> [ FAILED   ] 9 test(s), listed below:
> [ FAILED   ] Graph.ImageContainmentRelationship
> [ FAILED   ] LaplacianPyramid.GraphProcessing/0/randomInput/VX_BORDER_UNDEFINED/sz=128x128
> [ FAILED   ] LaplacianPyramid.GraphProcessing/1/randomInput/VX_BORDER_UNDEFINED/sz=256x256
> [ FAILED   ] LaplacianPyramid.GraphProcessing/2/randomInput/VX_BORDER_UNDEFINED/sz=640x480
> [ FAILED   ] LaplacianPyramid.GraphProcessing/18/lena/VX_BORDER_UNDEFINED
> [ FAILED   ] LaplacianPyramid.ImmediateProcessing/0/randomInput/VX_BORDER_UNDEFINED/sz=128x128
> [ FAILED   ] LaplacianPyramid.ImmediateProcessing/1/randomInput/VX_BORDER_UNDEFINED/sz=256x256
> [ FAILED   ] LaplacianPyramid.ImmediateProcessing/2/randomInput/VX_BORDER_UNDEFINED/sz=640x480
> [ FAILED   ] LaplacianPyramid.ImmediateProcessing/18/lena/VX_BORDER_UNDEFINED
> [ DISABLED ] 8190 test(s)
> ```
> 
> * HIP -
> 
> ```
> [ ======== ]
> [ ALL DONE ] 5820 test(s) from 69 test case(s) ran
> [ PASSED   ] 5796 test(s)
> [ FAILED   ] 24 test(s), listed below:
> [ FAILED   ] Graph.ImageContainmentRelationship
> [ FAILED   ] vxMapRemapPatch.MapRandomRemap
> [ FAILED   ] ColorConvert.OnRandomAndNatural/6/Immediate/RGB=>YUV4
> [ FAILED   ] ColorConvert.OnRandomAndNatural/7/Graph/RGB=>YUV4
> [ FAILED   ] ColorConvert.OnRandomAndNatural/14/Immediate/RGBX=>YUV4
> [ FAILED   ] ColorConvert.OnRandomAndNatural/15/Graph/RGBX=>YUV4
> [ FAILED   ] HarrisCorners.GraphProcessing/191/many_strong_corners/MIN_DISTANCE=30.0/SENSITIVITY=0.04/GRADIENT_SIZE=3/BLOCK_SIZE=7
> [ FAILED   ] HarrisCorners.GraphProcessing/194/many_strong_corners/MIN_DISTANCE=30.0/SENSITIVITY=0.04/GRADIENT_SIZE=5/BLOCK_SIZE=7
> [ FAILED   ] HarrisCorners.GraphProcessing/197/many_strong_corners/MIN_DISTANCE=30.0/SENSITIVITY=0.04/GRADIENT_SIZE=7/BLOCK_SIZE=7
> [ FAILED   ] HarrisCorners.GraphProcessing/200/many_strong_corners/MIN_DISTANCE=30.0/SENSITIVITY=0.10/GRADIENT_SIZE=3/BLOCK_SIZE=7
> [ FAILED   ] HarrisCorners.GraphProcessing/209/many_strong_corners/MIN_DISTANCE=30.0/SENSITIVITY=0.15/GRADIENT_SIZE=3/BLOCK_SIZE=7
> [ FAILED   ] HarrisCorners.ImmediateProcessing/191/many_strong_corners/MIN_DISTANCE=30.0/SENSITIVITY=0.04/GRADIENT_SIZE=3/BLOCK_SIZE=7
> [ FAILED   ] HarrisCorners.ImmediateProcessing/194/many_strong_corners/MIN_DISTANCE=30.0/SENSITIVITY=0.04/GRADIENT_SIZE=5/BLOCK_SIZE=7
> [ FAILED   ] HarrisCorners.ImmediateProcessing/197/many_strong_corners/MIN_DISTANCE=30.0/SENSITIVITY=0.04/GRADIENT_SIZE=7/BLOCK_SIZE=7
> [ FAILED   ] HarrisCorners.ImmediateProcessing/200/many_strong_corners/MIN_DISTANCE=30.0/SENSITIVITY=0.10/GRADIENT_SIZE=3/BLOCK_SIZE=7
> [ FAILED   ] HarrisCorners.ImmediateProcessing/209/many_strong_corners/MIN_DISTANCE=30.0/SENSITIVITY=0.15/GRADIENT_SIZE=3/BLOCK_SIZE=7
> [ FAILED   ] LaplacianPyramid.GraphProcessing/0/randomInput/VX_BORDER_UNDEFINED/sz=128x128
> [ FAILED   ] LaplacianPyramid.GraphProcessing/1/randomInput/VX_BORDER_UNDEFINED/sz=256x256
> [ FAILED   ] LaplacianPyramid.GraphProcessing/2/randomInput/VX_BORDER_UNDEFINED/sz=640x480
> [ FAILED   ] LaplacianPyramid.GraphProcessing/18/lena/VX_BORDER_UNDEFINED
> [ FAILED   ] LaplacianPyramid.ImmediateProcessing/0/randomInput/VX_BORDER_UNDEFINED/sz=128x128
> [ FAILED   ] LaplacianPyramid.ImmediateProcessing/1/randomInput/VX_BORDER_UNDEFINED/sz=256x256
> [ FAILED   ] LaplacianPyramid.ImmediateProcessing/2/randomInput/VX_BORDER_UNDEFINED/sz=640x480
> [ FAILED   ] LaplacianPyramid.ImmediateProcessing/18/lena/VX_BORDER_UNDEFINED
> [ DISABLED ] 8190 test(s)
> ```

@hansely @kiritigowda @asalmanp Master HIP only has the two Graph failures for Color. This introduces the two Immediate COlor Failures as well. Will need to look into that

**Review by asalmanp — APPROVED** [review]

**Review by kiritigowda — APPROVED** [review]

**Comment by kiritigowda** [comment]
@hansely did you check the 2 immediate node failures in HIP?

File: amd_openvx/openvx/api/vx_api.cpp (hunk 1 above)
Comment (rrawther) [reply amd_openvx/openvx/api/vx_api.cpp:947]: This is because of vxMapImagePatch which is copying from GPU to CPU. So input needs to be synced as well in the current implementation. Need to investigate this in future how to avoid this

**Review by rrawther — APPROVED** [review]
todo: need to investigate CreateImageFromHandle() for sync

**Comment by asalmanp** [comment]
> @hansely did you check the 2 immediate node failures in the HIP?

@kiritigowda @hansely @LakshmiKumar23 the two failures in **immediate node** color convert (RGB=>YUV4, RGBX=>YUV4) are the same as the **Graph** color convert (RGB=>YUV4, RGBX=>YUV4) failures . It is showing up in this branch because this PR pushes the immediate nodes to the GPU while on the master branch all immediate nodes, are running on CPU, so we don't see these failures. 
When we fix the Graph color convert (RGB=>YUV4, RGBX=>YUV4) issues, the corresponding immediate nodes failures for these formats will be fixed too.
